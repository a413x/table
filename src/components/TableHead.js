import React, { Component } from "react";
import PropTypes from 'prop-types';

import './styles/table.css';

class TableHead extends Component{
  constructor(props){
    super(props);
    this.state={
      activeCol: {field:'',order:'asc'}
    }
  }
  headerClick(field){
    let order = this.state.activeCol.order;
    if(this.state.activeCol.field == field){
      order = order == 'asc' ? 'desc' : 'asc';
    }else{
      order = 'asc';
    }
    this.setState({
      activeCol: {field:field,order:order}
    });
    this.props.sortFunc(field,order);
  }
  render(){
    const columns = this.props.columns;
    const active = this.state.activeCol;
    return(
      <thead>
        <tr>
          {columns.map((colObj,ind)=>
            <th scope = 'col' key={ind} onClick={()=>this.headerClick(colObj.field)}
              className = {active.field == colObj.field ? active.order : null}>
              {colObj.title}
            </th>
          )}
        </tr>
      </thead>
    );
  }
}

TableHead.propTypes = {
  columns: PropTypes.arrayOf(PropTypes.object).isRequired,
  sortFunc: PropTypes.func.isRequired
};

export default TableHead;
