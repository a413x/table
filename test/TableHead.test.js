import React from "react";
import {shallow,mount} from 'enzyme';

import TableHead from "../src/components/TableHead";
import data from './data.js';

describe('tests for TableHead component',() => {
  const initialState = {
    activeCol: {field:'',order:'asc'}
  };
  const tableColumns = data.tableColumns;
  const mockSortFunc = jest.fn();
  const props = {
    columns: tableColumns,
    sortFunc: mockSortFunc
  };

  const thContainer = shallow(<TableHead {...props} />);

  it('renders properly',()=>{
    expect(thContainer).toMatchSnapshot();
  });

  it('initialize with initial state',()=>{
    expect(thContainer.state()).toEqual(initialState);
  });

  describe('clicked on a th',()=>{
    const field = 'id';
    it('clicked once, headerClick should called, th should have class "asc"',()=>{
      thContainer.find('th').first().simulate('click');
      const currState = thContainer.state().activeCol;
      expect(currState).toEqual({field:field,order:'asc'});
      expect(mockSortFunc).toHaveBeenCalledWith(currState.field,currState.order);
      expect(thContainer.find('th').first().hasClass(currState.order)).toBe(true);
    });
    it('clicked twise, headerClick should called, th should have class "desc"',()=>{
      thContainer.find('th').first().simulate('click');
      const currState = thContainer.state().activeCol;
      expect(currState).toEqual({field:field,order:'desc'});
      expect(mockSortFunc).toHaveBeenCalledWith(currState.field,currState.order);
      expect(thContainer.find('th').first().hasClass(currState.order)).toBe(true);
    });
  });

  afterAll(() => {
    thContainer.setState(initialState);
  });
});
