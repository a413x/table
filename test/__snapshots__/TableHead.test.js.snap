// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for TableHead component renders properly 1`] = `
<thead>
  <tr>
    <th
      className={null}
      key="0"
      onClick={[Function]}
      scope="col"
    >
      #
    </th>
    <th
      className={null}
      key="1"
      onClick={[Function]}
      scope="col"
    >
      First name
    </th>
    <th
      className={null}
      key="2"
      onClick={[Function]}
      scope="col"
    >
      Last name
    </th>
    <th
      className={null}
      key="3"
      onClick={[Function]}
      scope="col"
    >
      E-mail
    </th>
    <th
      className={null}
      key="4"
      onClick={[Function]}
      scope="col"
    >
      Phone
    </th>
  </tr>
</thead>
`;
